flink-runtime/src/main/java/org/apache/flink/runtime/state/CompressibleFSDataOutputStream.java

```diff
@@ -41,6 +41,10 @@ public CompressibleFSDataOutputStream(
 
     @Override
     public long getPos() throws IOException {
+        // Underlying compression involves buffering, so the only way to report correct position is
+        // to flush the underlying stream. This lowers the effectivity of compression, but there is
+        // no other way, since the position is often used as a split point.
+        flush();
```

Comment:
Do you think it's the best way to fix it? I think this approach has a few drawbacks:

- As mentioned, lower efficiency of compression. Actually for some compression algorithms this may make it useless, if not harmful (if there is some significant amount of metadata that gets written on `flush`)
- It's somewhat strange that the result is different whether we call `getPos()` in the process of compressing the output, or we don't. This may later come as a big surprise for someone using `CompressibleFSDataOutputStream`.
- `Seek` operation in the corresponding `CompressibleFSDataInputStream` doesn't make sense for any position, only for some specific positions. Which is also not super bad, but confusing.

IMO, a better approach may be to change the semantics of `getPos()` for `CompressibleFSDataOutputStream`. This can represent the position in the **uncompressed** data.
Correspondingly, `seek` operation in `CompressibleFSDataInputStream` can be implemented having this semantics in mind: one can read the `compressingDelegate` until the desired position is reached.
This approach also doesn't have the problem of `seek` being valid for specific positions only, which is a nice side effect.

Reply:
Also, FYI, read operation in `SnappyFramedInputStream` also does the buffering:

```
@Override
public int read()
        throws IOException
{
    if (closed) {
        return -1;
    }
    if (!ensureBuffer()) {
        return -1;
    }
    return buffer[position++] & 0xFF;
}
```

So when we do `seek` in the `CompressibleFSDataInputStream` (which gets delegated to `delegate`) and then do `read` (which gets delegated to `compressingDelegate`) - I guess something weird can happen.

Reply:
There is a general rule that compressed blocks can't be split. This is why most systems distinguish between per-file, per-block or per-record compression.

In this case we're using per-record compression. It could be optimized by some form of record batching / block compression, but that would need to happen further up the stack, which goes beyond a bug fix.

---

> It's somewhat strange that the result is different whether we call getPos() in the process of compressing the output, or we don't. This may later come as a big surprise for someone using CompressibleFSDataOutputStream.

Not really. The position corresponds to the uncompressed stream. If you haven't flushed any data into it .. the position won't move.

> IMO, a better - albeit more involved - approach may be to change the semantics of getPos() for CompressibleFSDataOutputStream. This can represent the position in the uncompressed data.

`getPos()` has a strictly defined contract on the interface level, we need to hold up to it. Chaging it for a single implementation violates Liskov Substitution Principle.

```
  * <p>This method must report accurately report the current position of the stream. Various
     * components of the high-availability and recovery logic rely on the accurate
```

---

As I said, this could clearly be optimized by switching from a per-record compression to block compression, but that goes beyond critical bugfix and should be addressed separately (I'll be happy to do a review if you have any ideas).

WDYT?

Reply:
What are your thoughts on my second comment @dmvk? It's still an issue, isn't it?

Reply:
This issue is a bit more hidden, but as we stick with correct semantics of Stream interface should be addressed. I believe the following scenario should reveal it:

- Create `CompressibleFSDataInputStream` with snappy
- Seek to 1st item
- Read it (will be read correctly)
- Seek to say 5th item
- Read it (I believe it will be read incorrectly)

The main point is that the 2nd seek should be outside of the underlying buffer in the snappy stream.

Reply:
Your partially right. One missing point is that snappy always buffers at most until the next flush, so as long as we always read full record, it should work as expected. In case of partial reads, it works as you describe.

I've tried to fix the partial read case in https://github.com/apache/flink/pull/24079/commits/4a42f16a13f3d22c188a5020a8e8718394d11e0c, PTAL

Reply:
Great, thanks for additional testing!